staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -363,9 +374,8 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 	<-delayedStopCh
 	// wait for stoppedCh that is closed when the graceful termination (server.Shutdown) is finished.
 	<-stoppedCh
-
-	// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
-	s.HandlerChainWaitGroup.Wait()
+	// Wait for all requests in flight to drain, bounded by the RequestTimeout variable.
+	<-drainedCh
```

Comment:
I don't understand this change.

We wait for delayedStopCh 4 lines above. So the new goroutine you introduced doesn't add any new synchronization.

Can you explain?

Reply:
@wojtek-t  so this is the sequence of events:
- `A`:  T+0s: we receive `KILL` signal (`stopCh` is signaled)
- `B`: T+70s: `ShutdownDelayDuration` elapses, (`delayedStopCh` is signaled)
- `C`: T+70s: as soon as `ShutdownDelayDuration` elapses we should signal `WithWaitGroup` that it should send `5xx` with a `Retry-After` response header.
- `D`: wait for existing requests in flight to drain
- `E`: The `server.Shutdown` returns (`stoppedCh` is signaled)

So currently `C` happens after `E`, the new goroutine in this PR ensures that `C` happens immediately after `B`. 

> We wait for delayedStopCh 4 lines above.

yes, we don't really need that any longer, it can be simplified as shown below:
```
<-drainedCh
<-stoppedCh
```

Reply:
OK - that makes sense.

Please update this part to what you wrote above then.

Reply:
done, thanks!

also, as an FYI, have a WIP PR https://github.com/kubernetes/kubernetes/pull/101257 (i am going to split   that makes an attempt at:
- refactors the graceful termination logic and adds unit test so we can verify the expected behavior. 
- add a new mode of graceful termination to keep listening until the existing requests in flight drain

I am going to split it to two separate PRs - one with the refactor only and the other with new mode, I will ping you when it is ready for review.

Reply:
I've briefly skimmed that some time ago - splitting out refactoring part to a separate PR is a good thing to do.